### PR TITLE
[AAP-55913] Backport pagination fix PR #84 to stable-2.6

### PR DESCRIPTION
**Pull request by doziedotdev**
## Description

**Fixes**
- https://issues.redhat.com/browse/AAP-55913

**What is being changed?**
- Backporting the `build_url` fix from PR #84 to `stable-2.6`
- The function now strips scheme and hostname from fully-qualified URLs before processing

**Why is this change needed?**
- Customers on AAP 2.6 are experiencing 404 errors when `ansible.platform.gateway_api` lookup plugin fetches paginated results
- The `build_url` function incorrectly prepends `api/gateway/v1/` to fully-qualified URLs returned in pagination responses

**How does this change address the issue?**
- Strips the scheme and hostname from URLs starting with `http://` or `https://` before further processing
- Prevents malformed paths like `api/gateway/v1/https://hostname/api/...`

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Test update
- [ ] Refactoring (no functional changes)
- [ ] Development environment change
- [ ] Configuration change

## Self-Review Checklist
- [x] I have performed a self-review of my code
- [x] I have added relevant comments to complex code sections
- [x] I have updated documentation where needed
- [x] I have considered the security impact of these changes
- [x] I have considered performance implications
- [x] I have thought about error handling and edge cases
- [x] I have tested the changes in my local environment

### Prerequisites
- AAP 2.6 environment with Private Automation Hub

### Steps to Test
1. Run the following playbook:
```yaml
---
- name: test playbook
  hosts: localhost
  connection: local
  gather_facts: false
  vars:
    aap_validate_certs: false
    aap_hostname: <your-aap-host>
    aap_username: admin
    aap_password: <password>
  tasks:
    - name: "Get roles from Hub API (limit=5 forces pagination)"
      ansible.builtin.set_fact:
        hub_roles: "{{ query('ansible.platform.gateway_api', '/api/galaxy/pulp/api/v3/roles/',
                             query_params={'limit': 5},
                             host=aap_hostname, username=aap_username, password=aap_password,
                             verify_ssl=aap_validate_certs, return_all=true, max_objects=10000) }}"
    - name: "Show roles count"
      ansible.builtin.debug:
        msg: "Retrieved {{ hub_roles | length }} roles"
```

### Expected Results
- All pages of results should be fetched without 404 errors
- No malformed URLs like `api/gateway/v1/https://...` in error messages

### Tested Results
- **Without fix**: Failed with `api/gateway/v1/https://<host>/api/galaxy/pulp/api/v3/roles/?limit=5&offset=5` (404)
- **With fix**: Successfully retrieved 124 roles across multiple pages

## Additional Context

Backport of:
- #84 (SHA: 60f4705e0c0349dd34701694def4d4594426365a)

Original PR merged to `devel` on Dec 11, 2025. This backport brings the fix to customers on the stable-2.6 release line.

### Required Actions
- [ ] Requires documentation updates
- [ ] Requires downstream repository changes
  - Consider backporting to `stable-2.5` as well
- [ ] Requires infrastructure/deployment changes
- [ ] Requires coordination with other teams
- [ ] Blocked by PR/MR: #XXX

### CODE REVIEW

**Comment by antmaro** [comment]
Hi folks, anyone could please check this PR, maybe @KB-perByte or @john-westcott-iv . It seems that there were some checks failing during the pipeline, but they are already been fixed (not 100% sure) by the following PRs already merged:

- [x] https://github.com/ansible/ansible.platform/pull/120
- [x] https://github.com/ansible/ansible.platform/commit/446b2db2a85f50659f95956ce79efbccddbb5346#diff-20d1610925e550787a5f9dd9061f1e8ce9cee711b8f8165cd066b151179a9ba9

There are some customers, and also myself, waiting for this fix. So, I think if we force again the pipeline it should run successfully!

**Comment by KB-perByte** [comment]
Tagging @rohitthakur2590 for this!
He is the best person to help you out with reviews.

Thanks

**Comment by ivarmu** [comment]
Hi there! Is there anything I can do to get this PR merged?

**Comment by nickbhasin** [comment]
@web3dozie Can you fix the integration

**Comment by nickbhasin** [comment]
@web3dozie Can you fix the integration test for the same ?

File: plugins/module_utils/aap_module.py [hunk 1]
```diff
@@ -232,6 +232,9 @@ def warn(self, warning):
             super(AAPModule, self).warn(warning)
 
     def build_url(self, endpoint, query_params=None):
+        # Remove the host_url part if it is already present
+        if endpoint.startswith(("https://", "http://")):
+            endpoint = "/{0}".format('/'.join(endpoint.split('/')[3:]))
```
Comment (john-westcott-iv) [inline plugins/module_utils/aap_module.py]: **Suggestion (non-blocking):** Consider using `urlparse` (already imported on line 22) instead of string splitting for more robust URL handling. The split-by-`/`-and-take-from-index-3 approach works for standard `http(s)://host/path` URLs, but `urlparse` would handle edge cases (ports, encoded characters, etc.) more reliably and clearly expresses intent.

```suggestion
        if endpoint.startswith(("https://", "http://")):
            parsed = urlparse(endpoint)
            endpoint = parsed.path
            if parsed.query:
                endpoint = "{0}?{1}".format(endpoint, parsed.query)
```

That said, I understand this is a backport of an already-merged change, so maintaining parity with the `devel` branch implementation is reasonable.

File: changelogs/fragments/fix_pagination.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+bugfixes:
+  - Strip scheme and hostname from AAP url builder, which previously lead to malformed urls in the ansible.platform.gateway_api lookup plugin
```
Comment (john-westcott-iv) [inline changelogs/fragments/fix_pagination.yaml]: **Nit (non-blocking):** Minor grammar: "lead" should be "led" (past tense). The current phrasing reads "previously lead to malformed urls" but should be "previously led to malformed URLs".

```suggestion
  - Strip scheme and hostname from AAP url builder, which previously led to malformed URLs in the ansible.platform.gateway_api lookup plugin
```

**Review by john-westcott-iv — COMMENTED** [review]
## 📋 Code Review Summary

**Backport of the `build_url` pagination fix from PR #84 to `stable-2.6`.** The change correctly strips scheme and hostname from fully-qualified URLs before processing, preventing malformed paths like `api/gateway/v1/https://hostname/api/...` that caused 404 errors during pagination in the `gateway_api` lookup plugin.

**Files Reviewed**: 2 files
**Comments Posted**: 2 review comments

### 🔍 Issues Found
- 0 security/safety issues
- 0 correctness/logic issues
- 2 code quality suggestions (non-blocking)

### Overall Assessment
**LGTM** - The fix is correct and well-targeted. The pagination flow through `get_endpoint` → `build_url` is the right place for this fix, and the string-splitting approach works for the known use cases (standard HTTP(S) API pagination URLs). The changelog fragment is properly formatted.

### General Feedback
- The bug analysis is solid: `gateway_api` lookup plugin calls `get_endpoint(next_page)` with fully-qualified URLs from API responses, which routes through `build_url` and previously caused path mangling
- The fix is minimal and low-risk, appropriate for a stable branch backport
- Consider whether a backport to `stable-2.5` is also needed (mentioned in PR description)

**Review by john-westcott-iv — APPROVED** [review]
Approving with minor comments and pending test working.

**Comment by john-westcott-iv** [comment]
@rohitthakur2590 can you help get this integration test fixed? I'm not sure if its related to the change or not.

**Comment by doziedotdev** [comment]
Thanks for the review and the suggestions, @john-westcott-iv .  I've applied both (the `urlparse` refactor and the changelog grammar fix).

On the failing integration test: the `authenticator_maps_test` is failing at the "Assert that we can't create an incomplete authenticator_map" task. It looks like the test on this branch includes assertions for `team`, `organization`, and `role` validation errors (lines 62-64) that were carried over from the `devel` backport, but the Gateway API version that `stable-2.6` CI builds against only returns the `triggers` validation error, it rejects the request on that field alone before reaching the other validations.

The actual error from CI:
```
fatal: [testhost]: FAILED\! => {"changed": false, "msg": "{'triggers': ['Triggers must be a valid dict']}"}
```

The assertion then fails because it expects `'team': ['You must specify a team with the selected map type']` to also be present in `fail.msg`, which it isn't.

I believe the fix is to drop those three extra assertions so the test matches what `stable-2.6` actually returns, can you confirm that's the right approach? @john-westcott-iv @rohitthakur2590 Thanks.

File: plugins/module_utils/aap_module.py [hunk 3]
```diff
@@ -232,6 +232,13 @@ def warn(self, warning):
             super(AAPModule, self).warn(warning)
 
     def build_url(self, endpoint, query_params=None):
+        # Remove the host_url part if it is already present
+        if endpoint.startswith(("https://", "http://")):
+        if endpoint.startswith(("https://", "http://")):
```
Comment (djdanielsson) [inline plugins/module_utils/aap_module.py:236]: is this line supposed to be duplicated with the one above?

Comment (doziedotdev) [reply plugins/module_utils/aap_module.py:236]: amended the commit, thanks.

**Comment by djdanielsson** [comment]
integration test failed... not sure if it's actually related to this change or not. @KB-perByte @nickbhasin @john-westcott-iv any idea?

**Comment by ivarmu** [comment]
I think it's not related to the change:

```
TASK [authenticator_maps_test : Create Incomplete Authenticator Map] ***********
[ERROR]: Task failed: Module failed: {'triggers': ['Triggers must be a valid dict']}
Origin: /tmp/collections/ansible_collections/ansible/platform/tests/output/.tmp/integration/authenticator_maps_test-h5x3_1vp-ÅÑŚÌβŁÈ/tests/integration/targets/authenticator_maps_test/tasks/main.yml:49:7

47
48     # <Authenticator Maps> ------------------------
49     - name: Create Incomplete Authenticator Map
         ^ column 7

fatal: [testhost]: FAILED! => {"changed": false, "msg": "{'triggers': ['Triggers must be a valid dict']}"}
...ignoring

TASK [authenticator_maps_test : Assert that we can't create an incomplete authenticator_map] ***
[ERROR]: Task failed: Action failed: Assertion failed
Origin: /tmp/collections/ansible_collections/ansible/platform/tests/output/.tmp/integration/authenticator_maps_test-h5x3_1vp-ÅÑŚÌβŁÈ/tests/integration/targets/authenticator_maps_test/tasks/main.yml:57:7

55       ignore_errors: true
56
57     - name: Assert that we can't create an incomplete authenticator_map
         ^ column 7

fatal: [testhost]: FAILED! => {
    "assertion": "\"'team': ['You must specify a team with the selected map type']\" in fail.msg",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

It seems the Assert condition is not as complete as it should be.

**Comment by rohitthakur2590** [comment]
@web3dozie  looks like CI is passing for stable-2.6 branh  https://github.com/ansible/ansible.platform/pull/133 

The failing test comes from your branch having the devel version of the assertion (expects the team error), while stable-2.6’s Gateway only returns the triggers error.
 Rebase your branch on the latest stable-2.6. That will pull in the stable-2.6 test file, which already asserts the triggers message, so the integration test should pass without any extra changes.

**Comment by ivarmu** [comment]
@web3dozie can you rebase your branch to the stable-2.6 or merge stable-2.6 into your branch?

**Comment by doziedotdev** [comment]
I've rebased onto the latest stable-2.6

Thanks for the nudge @ivarmu  and @rohitthakur2590

**Comment by djdanielsson** [comment]
@rohitthakur2590 are you good with this merging?
